Modules/timemodule.c

```diff
@@ -2152,6 +2140,38 @@ pysleep(_PyTime_t secs)
         /* retry with the recomputed delay */
 #endif
     } while (1);
+#else
+    millisecs = _PyTime_AsMilliseconds(secs, _PyTime_ROUND_CEILING);
+    if (millisecs > LLONG_MAX) {
+        PyErr_SetString(PyExc_OverflowError,
+                        "sleep length is too large");
+        return -1;
+    }
+    
+    /* Allow sleep(0) to maintain win32 semantics, and as decreed
+     * by Guido, only the main thread can be interrupted.
+     */
+    ul_millis = (unsigned long)millisecs;
+    if (ul_millis == 0 || !_PyOS_IsMainThread()) {
+        Py_BEGIN_ALLOW_THREADS
+        Sleep(ul_millis);
+        Py_END_ALLOW_THREADS
+        return 0;
+    }
+    
+    Py_BEGIN_ALLOW_THREADS
+    rc = WaitForSingleObjectEx(hTimer, ul_millis, FALSE);
+    Py_END_ALLOW_THREADS
+    
+    if ((rc != WAIT_OBJECT_0) && (rc != WAIT_TIMEOUT)) {
+        return -1;
+    }
+
+    /* sleep was interrupted by SIGINT */
+    if (PyErr_CheckSignals()) {
+        return -1;
+    }
```

Comment:
The sleep should not be restarted if it was interrupted by a signal, and the signal handler does not raise an exception? https://www.python.org/dev/peps/pep-0475/ (the loop in the existing code)

Reply:
Old time.sleep() has this [same code](https://github.com/python/cpython/blob/3.10/Modules/timemodule.c#L2116) for Windows part, too. As i see in Windows imlementation the do {} while(1) loop probably useless because it will never go further to in 2nd cycle. It will succes the sleep or return -1 in any error checking. Was that a bad implementation in old code?

Reply:
A sleep can be interrupted by CTRL+C on Windows. If the Python signal handler of SIGINT does not raise an exception, time.sleep() must sleep again with the recomputed duration.

Reply:
Done